File: BNRSwiftJSON/BNRSwiftJSONTests/CardSet.m
//
//  CardSet.m
//  BNRSwiftJSON
//
//  Created by Zachary Waldowski on 5/20/15.
//  Copyright (c) 2015 Big Nerd Ranch Inc. All rights reserved.
//

#import "CardSet.h"
#import "Card.h"

@implementation CardSetObjC

+ (NSArray *)cardSetsFromDictionaries:(NSArray *)cardSetDicts
{
    NSMutableArray *cardSets = [[NSMutableArray alloc] initWithCapacity:cardSetDicts.count];
    for (NSDictionary *dict in cardSetDicts) {
        CardSetObjC *next = [[CardSetObjC alloc] initWithDictionary:dict];
        if (next == nil) {
            return nil;
        }
        [cardSets addObject:next];
    }
    return cardSets;
}

- (instancetype)initWithDictionary:(NSDictionary *)dict {
    self = [super init];
    if (!self) { return nil; }

    id(^forceUnwrap)(NSString *) = ^(NSString *key){
        id ret = dict[key];
        NSAssert(ret != nil, @"Unexpectedly found nil while fetching value for %@", key);
        return ret;
    };

    _name = forceUnwrap(@"name");
    _code = forceUnwrap(@"code");
    _gathererCode = dict[@"gathererCode"];
    _oldCode = dict[@"oldCode"];
    _magicCardsInfoCode = dict[@"magicCardsInfoCode"];

    ReleaseDateKind kind;
    NSDateComponents *comp = ReleaseDateFromString(dict[@"releaseDate"], &kind);
    NSAssert(kind == ReleaseDateKindFull, @"Unexpected date format");
    _releaseDate = comp.date;

    _border = forceUnwrap(@"border");
    _type = forceUnwrap(@"type");
    _block = dict[@"block"];
    if ([dict[@"onlineOnly"] boolValue]) {
        _onlineOnly = YES;
    }
    _booster = dict[@"booster"] ?: @[];

    NSArray *cardDicts = forceUnwrap(@"cards");
    NSMutableArray *cards = [[NSMutableArray alloc] initWithCapacity:cardDicts.count];
    for (NSDictionary *dict in cardDicts) {
        CardObjC *next = [[CardObjC alloc] initWithDictionary:dict];
        NSAssert(@"Unexpectedly found nil while converting value to %@", NSStringFromClass(CardObjC.self));
        [cards addObject:next];
    }
    _cards = cards;

    return self;

}

- (NSDictionary *)dictionaryValue {
    NSMutableDictionary *result = NSMutableDictionary.new;

    BOOL(^setIfNotNil)(NSString *, id) = ^(NSString *key, id value) {
        if (value == nil) {
            return NO;
        }
        result[key] = value;
        return YES;
    };

    BOOL(^setIfTrue)(NSString *, BOOL) = ^(NSString *key, BOOL value) {
        if (!value) {
            return NO;
        }
        result[key] = value ? @YES : @NO;
        return YES;
    };

    result[@"name"] = self.name;
    result[@"code"] = self.code;
    setIfNotNil(@"gathererCode", self.gathererCode);
    setIfNotNil(@"oldCode", self.oldCode);
    setIfNotNil(@"magicCardsInfoCode", self.magicCardsInfoCode);
    result[@"releaseDate"] = [ReleaseDateFullFormatter() stringFromDate:self.releaseDate];
    result[@"border"] = self.border;
    result[@"type"] = self.type;
    setIfNotNil(@"block", self.block);
    setIfTrue(@"onlineOnly", self.onlineOnly);
    if (self.booster.count != 0) {
        result[@"booster"] = self.booster;
    }

    NSMutableArray *cardDicts = [[NSMutableArray alloc] initWithCapacity:self.cards.count];
    for (CardObjC *card in self.cards) {
        [cardDicts addObject:card.dictionaryValue];
    }
    result[@"cards"] = cardDicts;

    return result;

}

@end

File: BNRSwiftJSON/BNRSwiftJSONTests/CardSet.swift
//
//  CardSet.swift
//  BNRSwiftJSON
//
//  Created by Zachary Waldowski on 5/19/15.
//  Copyright (c) 2015 Big Nerd Ranch Inc. Licensed under MIT.
//

import Foundation
import Argo
import Runes
import BNRSwiftJSON
import Result

struct CardSet {

    private(set) var name: String = ""
    private(set) var code: String = ""
    private(set) var gathererCode: String?
    private(set) var oldCode: String?
    private(set) var magicCardsInfoCode: String?
    private(set) var releaseDate: ReleaseDate
    private(set) var border: String = ""
    private(set) var type: String = ""
    private(set) var block: String?
    private(set) var isOnlineOnly: Bool = false
    private(set) var cards: [Card] = []

}

// MARK: - Argo-style

extension CardSet: Decodable {

    static func create(name: String)(_ code: String)(_ gathererCode: String?)(_ oldCode: String?)(_ magicCardsInfoCode: String?)(_ releaseDate: ReleaseDate)(_ border: String)(_ type: String)(_ block: String?)(_ isOnlineOnly: Bool?)(_ cards: [Card]) -> CardSet {
        return CardSet(name: name, code: code, gathererCode: gathererCode, oldCode: oldCode, magicCardsInfoCode: magicCardsInfoCode, releaseDate: releaseDate, border: border, type: type, block: block, isOnlineOnly: isOnlineOnly ?? false, cards: cards)
    }

    static func decode(j: Argo.JSON) -> Decoded<CardSet> {
        let f1 = CardSet.create
            <^> j <| "name"
            <*> j <| "code"
            <*> j <|? "gathererCode"
            <*> j <|? "oldCode"
            <*> j <|? "magicCardsInfoCode"
            <*> j <| "releaseDate"

        let f2 = f1
            <*> j <| "border"
            <*> j <| "type"
            <*> j <|? "block"
            <*> j <|? "onlineOnly"
            <*> j <|| "cards"

        return f2
    }

}

// MARK: - BNRSwiftJSON-style

extension CardSet {

    static func createWithJSON(json: BNRSwiftJSON.JSON) -> Result<CardSet> {
        let name = json["name"].string
        let code = json["code"].string
        let gathererCode = optional(json["gathererCode"].string)
        let oldCode = optional(json["oldCode"].string)
        let magicCardsInfoCode = optional(json["magicCardsInfoCode"].string)
        let releaseDate = json["releaseDate"].bind(ReleaseDate.createWithJSON)
        let border = json["border"].string
        let type = json["type"].string
        let block = optional(json["block"].string)
        let isOnlineOnly = fallback(json["onlineOnly"].bool, false)
        let cards = arrayOf(json["cards"]) as Result<[Card]>

        return bindAll(name, code, gathererCode, oldCode, magicCardsInfoCode, releaseDate, border, type) { (name, code, gathererCode, oldCode, magicCardsInfoCode, releaseDate, border, type) in
            mapAll(block, isOnlineOnly, cards) { (block, isOnlineOnly, cards) in
                CardSet(name: name, code: code, gathererCode: gathererCode, oldCode: oldCode, magicCardsInfoCode: magicCardsInfoCode, releaseDate: releaseDate, border: border, type: type, block: block, isOnlineOnly: isOnlineOnly, cards: cards)
            }
        }
    }

}
